File: Packages/RallyCore/Sources/RallyCore/Models/Sponsor.swift
import Foundation

/// Represents a sponsor with branded activations and content.
public struct Sponsor: Identifiable, Codable, Hashable, Sendable {
    public let id: String
    public let name: String
    public let logoURL: URL?
    public let websiteURL: URL?
    public let tier: SponsorTier

    public init(
        id: String,
        name: String,
        logoURL: URL? = nil,
        websiteURL: URL? = nil,
        tier: SponsorTier = .standard
    ) {
        self.id = id
        self.name = name
        self.logoURL = logoURL
        self.websiteURL = websiteURL
        self.tier = tier
    }
}

public enum SponsorTier: String, Codable, Hashable, Sendable {
    case presenting
    case premium
    case standard
}

/// Tracks a sponsor impression for analytics.
public struct SponsorImpression: Codable, Sendable {
    public let sponsorID: String
    public let placement: String
    public let activationID: String?
    public let eventID: String?
    public let timestamp: Date
    public let durationSeconds: TimeInterval?

    public init(
        sponsorID: String,
        placement: String,
        activationID: String? = nil,
        eventID: String? = nil,
        timestamp: Date = .now,
        durationSeconds: TimeInterval? = nil
    ) {
        self.sponsorID = sponsorID
        self.placement = placement
        self.activationID = activationID
        self.eventID = eventID
        self.timestamp = timestamp
        self.durationSeconds = durationSeconds
    }
}

File: Android/core/src/main/java/com/vanwagner/rally/core/model/Sponsor.kt
package com.vanwagner.rally.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Represents a sponsor with branded activations and content.
 */
@Serializable
data class Sponsor(
    val id: String,
    val name: String,
    val logoURL: String? = null,
    val websiteURL: String? = null,
    val tier: SponsorTier = SponsorTier.STANDARD,
    val isActive: Boolean = true
)

/**
 * Sponsorship level/tier.
 */
@Serializable
enum class SponsorTier {
    @SerialName("presenting") PRESENTING,
    @SerialName("premium") PREMIUM,
    @SerialName("standard") STANDARD;

    val displayName: String
        get() = name.lowercase().replaceFirstChar { it.uppercase() }
}

/**
 * Tracks a sponsor impression for analytics.
 */
@Serializable
data class SponsorImpression(
    val sponsorID: String,
    val placement: String,
    val activationID: String? = null,
    val eventID: String? = null,
    val timestamp: Long = System.currentTimeMillis(),
    val durationSeconds: Double? = null
)
